### ISSUE TITLE
Problem with corrupted data using allgather

### ISSUE THREAD

**Issue by jlmaccal** [mpi4py/mpi4py#186]
I'm having a problem with data corruption when using `allgather`, but only one of the HPC systems we use. I think the problem is very likely somewhere in the infiniband stack, but I'm at a complete loss to track it down.

We can trigger the error with the following simple test script:
```python
from mpi4py import MPI

NCOORD = 3600


class State:
    def __init__(self, coords):
        self.coords = coords


def main():
    world = MPI.COMM_WORLD
    rank = world.Get_rank()
    N = world.Get_size()

    with open(f"output_{rank}.txt", "w") as outfile:
        if rank == 0:
            # generate N random states
            states = []
            for i in range(N):
                coords = np.random.randn(NCOORD, 3)
                state = State(coords)
                states.append(state)

            # broadcast all states to each node
            world.bcast(states, root=0)

            # scatter a single state to each node
            state = world.scatter(states, root=0)

        else:
            states = world.bcast(None, root=0)
            state = world.scatter(None, root=0)

        while True:
            results = world.allgather(state)
            for i, (s1, s2) in enumerate(zip(states, results)):
                if (s1.coords != s2.coords).any():
                    print(f"Coordinates do not match on rank {rank}", file=outfile)
                    print(f"position {i}", file=outfile)
                    print("expected:", file=outfile)
                    print(s1.coords, file=outfile)
                    print("got:", file=outfile)
                    print(s2.coords, file=outfile)

                    mask = (s1.coords != s2.coords)
                    x1_diff = s1.coords[mask]
                    x2_diff = s2.coords[mask]
                    print("diff expected", file=outfile)
                    print(x1_diff, file=outfile)
                    print("diff got", file=outfile)
                    print(x2_diff, file=outfile)
                    raise ValueError()
            print("success")


if __name__ == "__main__":
    main()
```

This runs when ranks are confined to a single node, but fails whenever ranks span multiple nodes. The errors are either the `ValueError` triggered when the data does not match what is expected, an unknown pickle protocol error, or a pickle data truncated error. In each case, the data is apparently being corrupted.

Typically, one of the ranks on a node will complete successfully, whereas the remaining ranks will recieve garbage data. This makes me think this is some kind of data race.

The same script works fine on the other HPC system that we use, which makes me think it is a some kind of problem with the network stack, rather than with mpi4py. However, I have been getting pushback from our system adminstrators because the following equivalent c program runs without issue:

```c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>
#include <sys/time.h>

double **alloc_darray2d(int, int);
void free_darray2d(double **, int, int);
double ***alloc_darray3d(int, int, int);
void free_darray3d(double ***, int, int, int);
double ***alloc_parray2d(int, int);
void free_parray2d(double ***, int, int);
void gauss_rng(double *, int, double);

int main(int argc, char *argv[]) {
int i, j, p,  myid, numprocs, n, ntot, iter, found;
double *sbuf, *rbuf;
int ncoord = 3600;
int seed = -1, maxiter = 1000000;
unsigned int lseed;
double ***states, **state, ***results;
double sigma = 1.0;
const double randmax = (double)RAND_MAX + 1.0;
struct timeval curtime;
char fname[25];
FILE *fd;

   MPI_Init(NULL, NULL);
   MPI_Comm_rank(MPI_COMM_WORLD, &myid);
   MPI_Comm_size(MPI_COMM_WORLD, &numprocs);

   /* allocate 2d array of doubles state[ncoord][3] contiguously in memory */
   state = alloc_darray2d(ncoord, 3);
   /* allocate 3d array of doubles states[numprocs][ncoord][3] contiguously in memory */
   states = alloc_darray3d(numprocs, ncoord, 3);
   n = ncoord*3;
   ntot = n*numprocs;
   if (states == NULL) { fprintf(stderr, "error: failed to allocate states array\n"); }

   if (myid == 0) {
      /* configure the random number generator */
      if (argc >= 2) {
         seed = atoi(argv[1]);
         if (argc == 3) {
            maxiter = atoi(argv[2]);
         }
      }
      if (seed >= 0) {
         lseed = seed;
      } else {
         /* randomly seed rng using the current microseconds within the current second */
         gettimeofday(&curtime, NULL);
         lseed = curtime.tv_usec;
      }
      srandom(lseed);

      /* fill the states array with gaussian distributed random numbers */
      for (p = 0; p < numprocs; p++) {
         for (i = 0; i < ncoord; i++) {
            for (j = 0; j < 3; j++) {
               /* random returns integer between 0 and RAND_MAX, divide
                  by randmax gives number within [0,1) */
               states[p][i][j] = random()/randmax;
            }
         }
      }
      gauss_rng(&states[0][0][0], ntot, sigma);
   }
   /* scatter states array into state array at each process */
   MPI_Bcast(&states[0][0][0], ntot, MPI_DOUBLE, 0, MPI_COMM_WORLD);
   MPI_Scatter(&states[0][0][0], n, MPI_DOUBLE, &state[0][0], n, MPI_DOUBLE, 0, MPI_COMM_WORLD);
   /* gather state array from each process into results array */
   results = alloc_darray3d(numprocs, ncoord, 3);
   found = 0;
   for (iter = 1; iter <= maxiter; iter++) {
      MPI_Allgather(&state[0][0], n, MPI_DOUBLE, &results[0][0][0], n, MPI_DOUBLE, MPI_COMM_WORLD);
      snprintf(fname, 24, "scattergather.%d", myid);
      fd = fopen(fname, "w");
      /* compare states and results arrays */
      for (p = 0; p < numprocs; p++) {
         for (i = 0; i < ncoord; i++) {
            if (states[p][i][0] != results[p][i][0]
                || states[p][i][1] != results[p][i][1]
                || states[p][i][2] != results[p][i][2]) {
               fprintf(fd, "states and results differ on process %d at position %d:\n", myid, p);
               fprintf(fd, "states[%d][%d][0] = %g\nresults[%d][%d][0] = %g\n", p, i, p, i, states[p][i][0], results[p][i][0]);
               fprintf(fd, "states[%d][%d][1] = %g\nresults[%d][%d][1] = %g\n", p, i, p, i, states[p][i][1], results[p][i][1]);
               fprintf(fd, "states[%d][%d][2] = %g\nresults[%d][%d][2] = %g\n", p, i, p, i, states[p][i][2], results[p][i][2]);
               found = 1;
            }
         }
      }
      if (found == 1) { break; }
   }
   fclose(fd);
   MPI_Finalize();
}

#include <math.h>
void gauss_rng(double *z, int n, double sigma) {
/*
   gauss_rng converts an array z of length n filled with uniformly
   distributed random numbers in [0,1) into gaussian distributed
   random numbers with width sigma.
*/
double sq;
const double pi2 = 8.0*atan(1.0);
int i;

   if ((n/2)*2 != n) {
      fprintf(stderr, "error: n in gauss_rng must be even\n");
      return;
   }
   for (i = 0; i < n; i += 2) {
      sq = sigma*sqrt(-2.0*log(1.0 - z[i]));
      z[i] = sq*sin(pi2*z[i + 1]);
      z[i + 1] = sq*cos(pi2*z[i + 1]);
   }
}

/* allocate a double 2d array with subscript range
   dm[0,...,nx-1][0,...,ny-1]                      */
double **alloc_darray2d(int nx, int ny) {
  int i;
  double **dm;

  dm = (double **)malloc((size_t) nx*sizeof(double*));
  if (dm == NULL) { return NULL; }
  dm[0] = (double *)malloc((size_t) nx*ny*sizeof(double));
  if (dm[0] == NULL) {
     free((void *) dm);
     return NULL;
  }
  for (i = 1; i < nx; i++) dm[i] = dm[i - 1] + ny;
/* return pointer to array of pointers to rows */
  return dm;
}

void free_darray2d(double **dm, int nx, int ny) {
  free((void *) dm[0]);
  free((void *) dm);
}

/* allocate a 2d array of pointers to doubles with subscript range
   pm[0,...,nx-1][0,...,ny-1]           */
double ***alloc_parray2d(int nx, int ny) {
  int i;
  double ***pm;

  pm = (double ***)malloc((size_t) nx*sizeof(double**));
  if (pm == NULL) { return NULL; }
  pm[0]=(double **)malloc((size_t) nx*ny*sizeof(double*));
  if (pm[0] == NULL) {
     free((void *) pm);
     return NULL;
  }
  for (i = 1; i < nx; i++) pm[i] = pm[i - 1] + ny;
/* return pointer to array of pointers to rows */
  return pm;
}

void free_parray2d(double ***pm,int nx, int ny) {
  free((void *) pm[0]);
  free((void *) pm);
}

/* allocate 3d array of doubles with subscript range
   dm[0,...,nx-1][0,...,ny-1][0,...,nz-1]           */
double ***alloc_darray3d(int nx, int ny, int nz) {
  int i, j;
  double ***dm;

/* first, allocate 2d array of pointers to doubles */
  dm = alloc_parray2d(nx, ny);
  if (dm == NULL) return NULL;

/* allocate memory for the whole thing */
  dm[0][0] = (double *)malloc((size_t) nx*ny*nz*sizeof(double));
  if (dm[0][0] == NULL) {
    free_parray2d(dm, nx, ny);
    return NULL;
  }

/* set the pointers inside the matrix */
  for (i = 0; i < nx; i++) {
    for (j = 1; j < ny; j++) {
      dm[i][j] = dm[i][j - 1] + nz;
    }
    if (i < nx - 1) dm[i + 1][0] = dm[i][ny - 1] + nz;
  }

/* return pointer to array of matrix of pointers */
  return dm;
}

void free_darray3d(double ***dm, int nx, int ny, int nz) {
  free((void *) dm[0][0]);
  free_parray2d(dm, nx, ny);
```

**Comment by dalcinl** [comment]
1. Your Python code is missing a `import numpy as np` statement. Of course, this is a trivial omission and not the reason of your issues.
2. You did not mention what MPI implementation and version your HPC system is using. 
3. This you run `ldd` on the MPI extension module withing mpi4py to make sure the right libraries are linked in?
4. Note that your C code is not exactly equivalent to the Python one. You are using lowercase collectives like `allgather()`, mpi4py translates that to two successive calls to `MPI_Allgather()` and `MPI_Allgatherv()` ([see yourself](https://github.com/mpi4py/mpi4py/blob/e762c2f0a38aa98b758a02ea426e1ea28c76f4e2/src/mpi4py/MPI/msgpickle.pxi#L849). Perhaps you could try updating your C code to use vector collectives like `MPI_Allgatherv()` to try to reproduce. Or perhaps better: update your Python code to use uppercase collectives e.g. `Comm.Allgather()`, it is way more efficient for large data sets (but most likely not for the small arrays in your example).
5. This looks like an obvious issue in your MPI backend. Having parallel MPI code working perfectly well within a node, but not working across node boundaries is almost always an indication of a broken or misconfigured MPI library, or issues in the libraries related to the network (are you using special env variables to tweak MPI behavior?). Moreover, your code uses MPI collectives exclusively, and mpi4py's lowercase collectives in particular, in these cases your code eider works because it is right or hangs in a deadlock if you mess things up. On top of all that, you mention that the Python code works fine on a different HPC system.
6. Look at your MPI implementation documentation and see if you can change the network transport backed to plain slow TCP over Ethernet. Then test the multi-node configuration. If that works, then your sysadmin cannot keep ignoring you, otherwise he is in the verge of being unprofessional. BTW, remember your sysadmin that the C and Python codes are not truly equivalent, as I clarified in (3) above.
7. You can also try to ask your MPI backend (typically via env variables) to use different algorithms implementing the MPI collectives. Check the MPI docs to see if you have that option.

**Comment by dalcinl** [comment]
@jlmaccal Any update? Can we close this issue?

**Comment by jlmaccal** [comment]
Not a huge amount of progress on our end.

I suspect the issues are on our end, but I'm not sure how to troubleshoot or debug them. The issues really only seem to crop up when using mpi4py, rather than compiled C binaries. There seems to be some kind of config / library issue, but I'm at a loss for how to track it down.

I've done some more testing and we run into problems with multiple collective operations, not just allgather.

I've also tested using the tcp backend and the problems go away, backing your thoughts above.

**Comment by dalcinl** [comment]
Could you please try to run setting `export LD_PRELOAD=/path/to/libmpi.so` ?

Althernatively, could you build the [MPI-enabled Python interpreter](https://mpi4py.readthedocs.io/en/latest/appendix.html#mpi-enabled-python-interpreter) and use that binary executable to run your script? I'm wondering whether your issues are related to the way load extensions modules and the interaction with the MPI library. If that's the case, then using the MPI-enabled Python interpreter should fix your problems: the MPI library would load (with RTLD_GLOBAL) at the time Python starts running, rather than load (with RTLD_LOCAL) at the time mpi4py is imported.

Another thing to try: do `export MPI4PY_RC_THREAD_LEVEL=single` before running. This will request MPI initialization without multiple thread support.

BTW, you never mentioned what's the MPI implementation you are using.

**Comment by jlmaccal** [comment]
We're using openmpi 4.x.x (tried various versions), as well as intel mpi.

Using `export LD_PRELOAD=...` gives the following error:
```
--------------------------------------------------------------------------
An ORTE daemon has unexpectedly failed after launch and before
communicating back to mpirun. This could be caused by a number
of factors, including an inability to create a connection back
to mpirun due to a lack of common network interfaces and/or no
route found between them. Please check network connectivity
(including firewalls and network routing requirements).
--------------------------------------------------------------------------
```

I think I'm pointing to the correct `libmpi.so`, but the system has a fairly complicated set of lmod modules.

Setting `export MPI4PY_RC_THREAD_LEVEL=single` results in the same behaviour with corrupted data on some of the ranks.

**Comment by bartoldeman** [comment]
we have this issue as a ticket in the Compute Canada ticket system, and I've been debugging the issue (I'm working on the software stack).

In the end (and this can be reproduced with a small C code using just `MPI_Recv` and `MPI_Send`) there is a bug deep down in the software stack on an Omnipath-using cluster (Cedar), in either the `libpsm2.so` library or the kernel driver. Some more debugging needs to be done to see where exactly.

When the address you feed to `MPI_Recv` is not aligned on a 4 byte boundary, and you use the `MPI_BYTE` datatype, then *sometimes* the incoming data is misaligned, ie. shifted down to the 4 byte boundary.

Such a situation can easily happen with `Comm.allgather()` (since the pickled buffer in this case has 86599 bytes) but not `Comm.Allgather()` using MPI_INT or MPI_DOUBLE, which the above C example uses.

We also found that for some reason that isn't clear yet the combination of Open MPI 4.1.1 and libfabric 1.12.1 works ok, but all older Open MPI versions that I tested (4.0.3, 3.1.2, 2.1.1) and Intel MPI fail, with or without libfabric.

So this is not a bug in mpi4py. Of course mpi4py could work around it by aligning pickled buffers on 4-byte boundaries, but that's masking another bug.

I'll shortly open an issue at https://github.com/cornelisnetworks/opa-psm2 to see if we can get to the bottom of it there.

**Comment by dalcinl** [comment]
> Of course mpi4py could work around it by aligning pickled buffers on 4-byte boundaries

mpi4py cannot do that in general, as it does not have any control on how the `pickle` module produces its buffers not how Python allocates the `bytes` type. Of course, mpi4py could allocate an intermediate aligned buffer and copy the pickle output there, but it is an extra (potentially large) allocation and memory copy.

**Comment by bartoldeman** [comment]
what I meant was to add some padding in the receive buffer, by rounding up the displacements, but not the counts, to illustrate (untested!). But I wouldn't go here unless absolutely needed in any case!
```
diff --git a/src/mpi4py/MPI/msgpickle.pxi b/src/mpi4py/MPI/msgpickle.pxi
index 155f786..09544be 100644
--- a/src/mpi4py/MPI/msgpickle.pxi
+++ b/src/mpi4py/MPI/msgpickle.pxi
@@ -243,7 +243,7 @@ cdef object pickle_allocv(void **p, int n, MPI_Count cnt[], MPI_Aint dsp[]):
     cdef MPI_Count d=0
     for i from 0 <= i < n:
         dsp[i] = <MPI_Aint> d
-        d += cnt[i]
+        d += (cnt[i] + 3) & ~3;
     return pickle_alloc(p, d)
```

**Comment by dalcinl** [comment]
@bartoldeman Any updates?

**Comment by bartoldeman** [comment]
@dalcinl The above patch indeed works around the issue.
But in the end this is an issue lower down in the psm2 stack, which hasn't been fixed yet.

There are also other workarounds, as we've narrowed down the issue to only occur in rendez-vous mode, not in eager mode. So if you simply move up the eager-rendez-vous threshold up beyond your largest problematic case receive buffer size, e.g. `export PSM2_MQ_RNDV_HFI_THRESH=1048576`, the problem goes away (with some hopefully minor performance implications).

I think this issue can be closed then. It can stay here archived for reference.

**Comment by bartoldeman** [comment]
@dalcinl and just a day later the issue at https://github.com/cornelisnetworks/opa-psm2/issues/64 was fixed. So once `libpsm2.so` will be updated to the next release this problem will go away.

**Comment by BrendanCunningham** [comment]
> @dalcinl The above patch indeed works around the issue. But in the end this is an issue lower down in the psm2 stack, which hasn't been fixed yet.
> 
> There are also other workarounds, as we've narrowed down the issue to only occur in rendez-vous mode, not in eager mode. So if you simply move up the eager-rendez-vous threshold up beyond your largest problematic case receive buffer size, e.g. `export PSM2_MQ_RNDV_HFI_THRESH=1048576`, the problem goes away (with some hopefully minor performance implications).
> 
> I think this issue can be closed then. It can stay here archived for reference.

We did find and fix an issue, the issue we believe, in the PSM2 stack.

Are you still seeing the issue even after the fix on opa-psm2#64?
